aicsimageio/aics_image.py

```diff
@@ -0,0 +1,418 @@
+        ----------
+        image: types.ImageLike
+            A string, Path, fsspec based file, or, numpy or dask array, to read.
+        known_dims: Optional[str]
+            Optional string with the known dimension order. If None, the reader will
+            attempt to parse dim order.
+        kwargs: Any
+            Extra keyword arguments that will be passed down to the reader subclass.
+
+        Examples
+        --------
+        Initialize an image and read the slices specified as a numpy array.
+
+        >>> img = AICSImage("my_file.tiff")
+        ... zstack_t8 = img.get_image_data("ZYX", S=0, T=8, C=0)
+
+        Initialize an image, construct a delayed dask array for certain slices, then
+        read the data.
+
+        >>> img = AICSImage("my_file.czi")
+        ... zstack_t8 = img.get_image_dask_data("ZYX", S=0, T=8, C=0)
+        ... zstack_t8_data = zstack_t8.compute()
+
+        Initialize an image with a dask or numpy array.
+
+        >>> data = np.random.rand(100, 100)
+        ... img = AICSImage(data)
+
+        Initialize an image from S3 with s3fs.
+
+        >>> fs = s3fs.S3FileSystem("s3://my_bucket")
+        ... f = fs.open("s3://my_bucket/my_file.tiff")
+        ... img = AICSImage(f)
+
+        Initialize an image and pass arguments to the reader using kwargs.
+
+        >>> img = AICSImage("my_file.czi", chunk_by_dims=["T", "Y", "X"])
+
+        Notes
+        -----
+        Constructor for AICSImage class intended for providing a unified interface for
+        dealing with microscopy images. To extend support to a new reader simply add a
+        new reader child class of Reader ([readers/reader.py]) and add the class to
+        SUPPORTED_READERS variable.
+        """
+        pass
+
+    @staticmethod
+    def determine_reader(image: types.ImageLike) -> Reader:
+        """
+        Cheaply check to see if a given file is a recognized type and return the
+        appropriate reader for the file.
+        """
+        pass
+
+    @property
+    def reader(self) -> Reader:
+        """
+        This property returns the object created to read the image file type.
+        The intent is that if the AICSImage class doesn't provide a raw enough
+        interface then the base class can be used directly.
+        """
+        pass
+
+    @property
+    def dask_data(self) -> da.Array:
+        """
+        Returns the image as a dask array with dimension ordering "STCZYX".
+        """
+        pass
+
+    @property
+    def data(self) -> np.ndarray:
+        """
+        Return the image as a numpy array with dimension ordering "STCZYX".
+        """
+        pass
+
+    @property
+    def dtype(self) -> np.dtype:
+        """
+        Data-type of the image array's elements.
+        """
+        pass
+
+    @property
+    def shape(self) -> Tuple[int]:
+        """
+        Tuple of the image array's dimensions.
+        """
+        pass
+
+    @property
+    def dims(self) -> Dimensions:
```

Comment:
I guess I still am not convinced of the use case for this, when we have shape also.  An aicsImage is always a STCZYX and so the dimension string names are already known.

Reply:
Would you prefer `myAicsImage.dims.Time` or `myAicsImage.dims.T`?

Reply:
For me, as a user, I know what STCZYX all mean and I'm 100% used to those abbreviations.  The only odd one is S.  In my mind this is replacing what I would have done before by calling `myimage.size_t` now I will do with `myimage.dims.T` which feels like a perfectly good tradeoff.

Reply:
I think I agree. I had left the comment in the dimensions.py file to exactly discuss the resolution of users confusion about dimension naming by making each dimension its own named tuple.

So `myAicsImage.dims.S.size == {int}` and `myAicsImage.dims.name == "Scene"` for example? But I think that is just verbose when we could have documentation. Thoughts though?

Reply:
My opinion: I believe both the readers and the aicsimage should use dimension names or abbreviations that are "native".  Like if I am using CZIReader and i get a "B" , that's czi's naming for it.  Just like it's up to czi to document what it means, it's our job to document what our STCZYX mean.

Reply:
Agree.